### DISCUSSION TITLE
Possible to do a ControlOut request with a timeout

### DISCUSSION THREAD

**Discussion by andrewdavidmackenzie** [kevinmehall/nusb#91] [Q&A]
Great crate, easy to use, thanks for that!

I'm using nusb from my Iced application, and using all the async methods.
Is it possible to do an Interface::control_out() call, but with a timeout?
If not, I guess I will find a way to cancel the future after some time, but as I saw it on control_out_blocked() I thought I'd ask.
thanks

**Comment by martinling** [comment]
Issue #4 is open for this feature, but it would currently require either the `futures_lite` crate to add a timeout API, or nusb to gain additional dependencies.

**Comment by andrewdavidmackenzie** [comment]
Thanks. Although the issue title is a bit misleading no?

It would be "timeouts for async/non-blocking IO" no?

I guess I'll implement the "race", using select I guess?

**Reply by kevinmehall** [reply]
Yes, `select` or `race` with an [`async_io::Timer`](https://docs.rs/async-io/latest/async_io/struct.Timer.html), or [`tokio::time::timeout`](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html) are the way to go there. You need a more full-fledged async runtime to give a future a timeout.

**Comment by martinling** [comment]
Yeah the title is misleading at this point I guess.

For some context, when that issue was opened, the `control_in_blocking` and `control_out_blocking` methods hadn't been added yet.

But those timeouts aren't really integrated into the async side of things - those methods just use the synchronous APIs on the OS. And those methods are only available for control transfers.

So having a `block_on` with a timeout, or a `TransferFuture` with a built-in timeout, would still be really useful.

**Reply by kevinmehall** [reply]
The proposal in #4 is to integrate the  `block_on` mini-executor and implement the timeout there when blocking until a completion event. That's easier to do than making the Future itself time out because the underlying CondVar used for waiting already implements a timeout, so you basically use the blocked thread as the timer. 

The 3 operating systems support timeouts for transfers in different ways, making it hard to abstract over any underlying OS functionality at the USB level:
* macOS supports async transfer timeouts natively.
* Windows has a timeout that is set per-endpoint, not per-transfer.
* Linux has no timeout support in its async APIs and expect you to just cancel the transfer after your own timeout.

To make the `Future` itself time out without the cooperation of the executor or OS-level USB API is trickier and would probably require depending on [`async_io::Timer`](https://docs.rs/async-io/latest/async_io/struct.Timer.html) or Tokio to have a background thread emitting timer events. That's essentially the same as what you can do externally with a timer implementation of your choice and  `select`.

It seems that typically Rust async IO futures such as those provided by `tokio` and `async_io` don't time out on their own, and expect you to race with a timer future, or use a [convenience wrapper](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html) that does that for you.

A user of a blocking API can't implement a timeout externally, and may not be familiar with the async ecosystem to do it themselves, so it seems like something the library should provide there, but I'm not sure it belongs in nusb's async APIs for those reasons.
